### PR TITLE
add docs for the kubernetes adapter

### PR DESCRIPTION
**Pull request by johrstrom**
https://osc.github.io/ood-documentation-test/add-k8s-docs/

Adds documentation for the kubernetes adapter. Some TODOs in there.

### CODE REVIEW

**Comment by treydock** [comment]
Some other TODOs would be documenting job-pod-reaper and k8-namespace-reaper since those are key to how we use Kubernetes with OnDemand.  If we document our Kyverno policies then also need to document k8-ldap-configmap which is where Kyverno knows what UID/GID/groups are allowed.  I can handle those things if you want to just add appropriate place via TODO items.

Also I would recommend along with k8s-jupyter documentation to also document how we currently do bc_osc_jupyter since that's a simpler Pod setup and something I think other sites would very much like to mimic, treating K8 pods like compute nodes.

Also FYI the OSC custom Kyverno policies are here: https://github.com/OSC/osc-helm-charts/tree/main/charts/kyverno-policies/templates.  That chart also depends on Kyverno upstream policies that also work with OnDemand.

File: source/installation/resource-manager/kubernetes.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+to use prefix ``user-``, then the namespace would be ``user-oakley``.
+
+The networking policy ensures that pods cannot communicate inbetween namespaces.
+
+The rolebindings bind the user, ``oakley`` in this case, to the ``ood-user``
+ClusterRole.  It also creates a service account in the namespace ``user-oakley``.
+The service account is used in initializing configmaps and secrets for the Open
+OnDemand pods.
+
+**TODO** Need to find, share and document the ood-user ClusteRole.
```
Comment (treydock) [inline source/installation/resource-manager/kubernetes.rst]: https://github.com/OSC/ondemand/blob/master/hooks/k8s-bootstrap/ondemand.yaml

File: source/installation/resource-manager/kubernetes.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  # /etc/ood/config/clusters.d/my_k8s_cluster.yml
+  ---
+  v2:
+    job:
+      auth:
+        type: 'managed'
+
+        type: 'odic'
+
+        type: 'gke'
```
Comment (treydock) [inline source/installation/resource-manager/kubernetes.rst]: Is it intentional to have all 3 types listed? Maybe clearer to document the types outside of the YAML example

**Comment by treydock** [comment]
Also be good to document how to pass GPUs:

```yaml
script:
  gpus_per_node: 1
```

The part OSC doesn't use is how to configure something besides NVIDIA, this is default:

```yaml
script:
  native:
    gpu_type: nvidia.com/gpu
```

File: source/installation/resource-manager/kubernetes.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+fsGroup is always the same as runAsGroup. runAsNonRoot is always set to true.
+supplementalGroups are empty by default. One can automatically populate them with a
+cluster configuration above or specify them for each app individually.
+
+You should have policies in place to enforce these.
+
+Bootstrapping an Open OnDemand Kuberenetes cluster
+**************************************************
+
+We ship with `open ondemand provided hooks`_ to bootstrap users when the login
```
Comment (treydock) [inline source/installation/resource-manager/kubernetes.rst:100]: Might be good to show example of how pre-hook.sh: https://github.com/OSC/osc-ood-config/blob/master/hooks/pre-hook.sh

Also how we configure the hook:

```
# cat /etc/ood/config/hook.env
CLIENT_ID="<OIDC Client ID>"
CLIENT_SECRET="<OIDC Client secre>"
K8S_USERNAME_PREFIX="dev-"
NAMESPACE_PREFIX="user-"
NETWORK_POLICY_ALLOW_CIDR="<CIDR network range for OnDemand web node network>"
IDP_ISSUER_URL="https://idp-dev.osc.edu/auth/realms/osc"
IMAGE_PULL_SECRET="private-docker-registry"
REGISTRY_DOCKER_CONFIG_JSON="/opt/osc/etc/docker-registry-auth/docker-registry.osc.edu-robot-ondemand-read.json"
```

**Comment by treydock** [comment]
Be good to call out ood_portal.yml changes:

```yaml
pun_pre_hook_root_cmd: "/opt/ood-apps-config/hooks/pre-hook.sh"
pun_pre_hook_exports: OIDC_ID_TOKEN,OIDC_ACCESS_TOKEN,OIDC_REFRESH_TOKEN
```

**Comment by treydock** [comment]
Somewhere we may also need to call out how we configured Keycloak to have OnDemand audience be valid for Kubernetes so the OnDemand OAuth token is also seen as valid for a different OIDC client for Kubernetes.

This is done through Keycloak with oidc-audience-mapper protocol mapper.  Here is relevant JSON associated to OnDemand client, I can also get screenshots:

```json
  "protocolMappers" : [ {
    "id" : "2d16810f-b8bc-5d19-adc0-41c0b7632ee7",
    "name" : "kubernetes",
    "protocol" : "openid-connect",
    "protocolMapper" : "oidc-audience-mapper",
    "consentRequired" : false,
    "config" : {
      "included.client.audience" : "kubernetes",
      "id.token.claim" : "true",
      "access.token.claim" : "true"
    }
  } ],
```

File: source/installation/resource-manager/kubernetes.rst (hunk 2 above)
Comment (johrstrom) [reply source/installation/resource-manager/kubernetes.rst]: Yea kindof, only gke takes additional things besides type. But yea, we cna fix that up to be simpler.

File: source/installation/resource-manager/kubernetes.rst (hunk 3 above)
Comment (johrstrom) [reply source/installation/resource-manager/kubernetes.rst:100]: Thanks, added this comment to #439 which we'll take care of this week.

File: source/installation/resource-manager/kubernetes.rst (hunk 1 above)
Comment (johrstrom) [reply source/installation/resource-manager/kubernetes.rst]: I think I fixed this, reworded it a bit to fit around the link.

File: source/installation/resource-manager/kubernetes.rst (hunk 2 above)
Comment (johrstrom) [reply source/installation/resource-manager/kubernetes.rst]: This is done, I just gave them all code blocks.

**Comment by johrstrom** [comment]
I'm going to merge this now because we're announcing at Gateways tomorrow. There are spots for missing items.  Looks like we'll need to start breaking this up into smaller pages.

**Pull request merged**